lib/backend/k8s/resources/ipam_block.go

```diff
@@ -0,0 +1,268 @@
+	"github.com/projectcalico/libcalico-go/lib/net"
+	log "github.com/sirupsen/logrus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+)
+
+const (
+	IPAMBlockResourceName = "IPAMBlocks"
+	IPAMBlockCRDName      = "ipamblocks.crd.projectcalico.org"
+)
+
+func NewIPAMBlockClient(c *kubernetes.Clientset, r *rest.RESTClient) K8sResourceClient {
+	// Create a resource client which manages k8s CRDs.
+	rc := customK8sResourceClient{
+		clientSet:       c,
+		restClient:      r,
+		name:            IPAMBlockCRDName,
+		resource:        IPAMBlockResourceName,
+		description:     "Calico IPAM blocks",
+		k8sResourceType: reflect.TypeOf(apiv3.IPAMBlock{}),
+		k8sResourceTypeMeta: metav1.TypeMeta{
+			Kind:       apiv3.KindIPAMBlock,
+			APIVersion: apiv3.GroupVersionCurrent,
+		},
+		k8sListType:  reflect.TypeOf(apiv3.IPAMBlockList{}),
+		resourceKind: apiv3.KindIPAMBlock,
+	}
+
+	return &ipamBlockClient{rc: rc}
+}
+
+// Implements the api.Client interface for IPAMBlocks.
+type ipamBlockClient struct {
+	rc customK8sResourceClient
+}
+
+func (c ipamBlockClient) toV1(kvpv3 *model.KVPair) (*model.KVPair, error) {
+	cidrStr := kvpv3.Value.(*apiv3.IPAMBlock).Annotations["projectcalico.org/cidr"]
+	_, cidr, err := net.ParseCIDR(cidrStr)
+	if err != nil {
+		return nil, err
+	}
+
+	ab := kvpv3.Value.(*apiv3.IPAMBlock)
+
+	// The CRD needs ints, not *int, so convert here.
+	allocations := []*int{}
+	for _, a := range ab.Spec.Allocations {
+		if a == -1 {
+			allocations = append(allocations, nil)
+		} else {
+			a2 := a
+			allocations = append(allocations, &a2)
+		}
+	}
+
+	// Convert attributes.
+	attrs := []model.AllocationAttribute{}
+	for _, a := range ab.Spec.Attributes {
+		attrs = append(attrs, model.AllocationAttribute{
+			AttrPrimary:   a.AttrPrimary,
+			AttrSecondary: a.AttrSecondary,
+		})
+	}
+
+	return &model.KVPair{
+		Key: model.BlockKey{
+			CIDR: *cidr,
+			UID:  &ab.UID,
+		},
+		Value: &model.AllocationBlock{
+			CIDR:           *cidr,
+			Affinity:       ab.Spec.Affinity,
+			StrictAffinity: ab.Spec.StrictAffinity,
+			Allocations:    allocations,
+			Unallocated:    ab.Spec.Unallocated,
+			Attributes:     attrs,
+			Deleting:       ab.Spec.Deleting,
+		},
+		Revision: kvpv3.Revision,
+	}, nil
+}
+
+func (c ipamBlockClient) v3Fields(k model.Key) (name, cidr string) {
+	// Name is calculated using the CIDR, replacing characters which
+	// are not allowed in the Kubernetes API.
+	// e.g., 10.0.0.1/26 -> 10-0-0-1-26
+	cidr = fmt.Sprintf("%s", k.(model.BlockKey).CIDR)
+	name = strings.Replace(cidr, ".", "-", -1)
+	name = strings.Replace(name, ":", "-", -1)
+	name = strings.Replace(name, "/", "-", -1)
+	return
+}
+
+func (c ipamBlockClient) toV3(kvpv1 *model.KVPair) *model.KVPair {
+	name, cidr := c.v3Fields(kvpv1.Key)
+
+	ab := kvpv1.Value.(*model.AllocationBlock)
+
+	// The CRD needs ints, not *int, so convert here.
```

Comment:
For my education, what does happen to `*int`?

Reply:
Ooooh, this might have been fixed since I prototyped this.  Here was the original issue: https://github.com/kubernetes/gengo/issues/11

Reply:
That issue is for `*[]int`, not `[]*int`?

Reply:
Bah, you're right.

Let me just try using pointers and see what happens.. it has been a while since I prototyped this.

Reply:
Huh, this looks like it's working now.

I swore this was broken when I did this way back. Guess it's been fixed!